Sources/NIOExtras/HTTP1ProxyConnectHandler.swift

```diff
@@ -0,0 +1,299 @@
+
+        case .failed, .completed:
+            break
+        }
+        context.fireChannelInactive()
+    }
+
+    public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        preconditionFailure("We don't support outgoing traffic during HTTP Proxy update.")
+    }
+
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        switch self.unwrapInboundIn(data) {
+        case .head(let head):
+            self.handleHTTPHeadReceived(head, context: context)
+        case .body:
+            self.handleHTTPBodyReceived(context: context)
+        case .end:
+            self.handleHTTPEndReceived(context: context)
+        }
+    }
+
+    private func sendConnect(context: ChannelHandlerContext) {
+        guard case .initialized = self.state else {
+            // we might run into this handler twice, once in handlerAdded and once in channelActive.
+            return
+        }
+
+        let timeout = context.eventLoop.scheduleTask(deadline: self.deadline) {
+            switch self.state {
+            case .initialized:
+                preconditionFailure("How can we have a scheduled timeout, if the connection is not even up?")
+
+            case .connectSent, .headReceived:
+                self.failWithError(Error.httpProxyHandshakeTimeout, context: context)
+
+            case .failed, .completed:
+                break
+            }
+        }
+
+        self.state = .connectSent(timeout)
+
+        let head = HTTPRequestHead(
+            version: .init(major: 1, minor: 1),
+            method: .CONNECT,
+            uri: "\(self.targetHost):\(self.targetPort)",
+            headers: self.headers
+        )
+
+        context.write(self.wrapOutboundOut(.head(head)), promise: nil)
+        context.write(self.wrapOutboundOut(.end(nil)), promise: nil)
+        context.flush()
+    }
+
+    private func handleHTTPHeadReceived(_ head: HTTPResponseHead, context: ChannelHandlerContext) {
+        switch self.state {
+        case .connectSent(let scheduled):
+            switch head.status.code {
+            case 200..<300:
+                // Any 2xx (Successful) response indicates that the sender (and all
+                // inbound proxies) will switch to tunnel mode immediately after the
+                // blank line that concludes the successful response's header section
+                self.state = .headReceived(scheduled)
+            case 407:
+                self.failWithError(Error.proxyAuthenticationRequired, context: context)
+
+            default:
+                // Any response other than a successful response indicates that the tunnel
+                // has not yet been formed and that the connection remains governed by HTTP.
+                self.failWithError(Error.invalidProxyResponseHead(head), context: context)
+            }
+        case .failed:
+            break
+        case .initialized, .headReceived, .completed:
```

Comment:
@Lukasa how much does the `HTTPDecoder` validate state for us? Presumably a remote peer could arrange for us to hit `.completed` state quite easily. Possibly the `.headReceived` state too depending on what the decoder does.

Reply:
Quite a lot, it won't emit these parts in the wrong order.